### DISCUSSION TITLE
How to enable python selectors with pytest on linux?

### DISCUSSION THREAD

**Discussion by AyushExel** [pytest-dev/pytest#9010] [Q&A]
I have a working solution for user input timeout using `selectors` in linux/uninx and `msvcrt` in windows. I have written a test that works with windows version but throws `PermissionError: [Errno 1] Operation not permitted with pytest on linux`.

**stdin_timeout**

```
def posix_stdin_timeout(prompt="", timeout=DEFAULT_TIMEOUT, timeout_log=""):
    echo(prompt)
    sel = selectors.DefaultSelector()
    sel.register(sys.stdin, selectors.EVENT_READ)
    keys_and_events = sel.select(timeout)

    if keys_and_events:
        key, _ = keys_and_events[0]
        return key.fileobj.readline().rstrip(LF)
    else:
        echo(LF)
        termios.tcflush(sys.stdin, termios.TCIFLUSH)
        raise InputTimeoutError(timeout_log)
```

**test**
```
def test_timeout(pytestconfig):
    capmanager = pytestconfig.pluginmanager.getplugin("capturemanager")
    capmanager.suspend_global_capture(in_=True)

    timeout_log = "input timeout!"
    try:
        stdin_timeout("waiting for input", 1, timeout_log)
    except InputTimeoutError as e:
        assert e.message == timeout_log
    capmanager.resume_global_capture()
```
These functions work fine when running as python applications but running the above test with pytest throws this error:
```

self = <selectors.EpollSelector object at 0x7f0e8d116390>, fileobj = <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>, events = 1, data = None

    def register(self, fileobj, events, data=None):
        key = super().register(fileobj, events, data)
        poller_events = 0
        if events & EVENT_READ:
            poller_events |= self._EVENT_READ
        if events & EVENT_WRITE:
            poller_events |= self._EVENT_WRITE
        try:
            self._selector.register(key.fd, poller_events)
E           PermissionError: [Errno 1] Operation not permitted

/home/.pyenv/versions/3.7.11/lib/python3.7/selectors.py:359: PermissionError
```
Is there a way I can make this test succeed?

**Comment by RonnyPfannschmidt** [comment]
selectors is directly incompatible for stdio capture - doesntit have a test mode?

**Comment by initial-d** [comment]
`suspend_global_capture(in_=True)` is not enough to make this a good unit test target on Linux. Under pytest, `stdin` is often not a normal interactive terminal fd, and `epoll`/`selectors` can fail on the object pytest has installed for capture. So the failure is about the test environment, not necessarily about your application code.

I would test this in two layers:

1. **Unit-test the timeout logic without real stdin**

   Put the selector creation behind a small seam, or pass in the file object/selector. Then test the branches with a fake selector:

```python
def posix_stdin_timeout(prompt="", timeout=DEFAULT_TIMEOUT, timeout_log="", selector_factory=selectors.DefaultSelector, stdin=sys.stdin):
    echo(prompt)
    sel = selector_factory()
    sel.register(stdin, selectors.EVENT_READ)
    keys_and_events = sel.select(timeout)
    ...
```

Then your unit test can make `select()` return `[]` and assert `InputTimeoutError` without touching pytest's captured stdin at all.

2. **Integration-test the real POSIX stdin behavior in a subprocess/pty**

   If you really want to prove the Linux terminal behavior, run a tiny script under a pseudo-terminal using `pty`/`pexpect` or a subprocess with controlled stdin. That gives your code a real fd instead of pytest's capture object.

For example, the high-level shape is:

```python
def test_timeout_logic_with_fake_selector():
    class FakeSelector:
        def register(self, fileobj, events):
            pass
        def select(self, timeout):
            return []

    with pytest.raises(InputTimeoutError):
        posix_stdin_timeout(timeout=0.01, selector_factory=FakeSelector, stdin=object())
```

The direct test against `sys.stdin` inside pytest will always be brittle because pytest intentionally owns stdio during capture. I would keep the real selector test outside capture, preferably in an integration test that creates its own terminal-like input source.

**Comment by ShreeChaturvedi** [comment]
`selectors.DefaultSelector` is `EpollSelector` on Linux, and epoll rejects regular files and /dev/null with exactly that EPERM. `suspend_global_capture(in_=True)` puts fd 0 back to whatever pytest itself was launched with. If that wasn't a terminal, you still get the error.

```
pytest -s test_stdin.py
```

from a real shell and the `register` goes through.

`SelectSelector` will accept those fds, but `select()` calls /dev/null readable instantly, so your timeout branch never fires. Gate the test on `os.isatty(0)` if it has to run in CI too.
